File: leetcode/java/gasstation.java
/*
 * there are N gas station along a circular route, where the amount of gas at station i is gas[i]
 * you have a car with an unlimited gas tank and it costs cost[i] of gas to travel from station i to its next station (i+1).
 * you begin the journey with an empty tank at one of the gas stations.
 * return the starting gas station's index if you can travel around the circuit once, otherwise return -1.
 * */
import java.io.*;
import java.util.*;

public class gasstation{
    public int completeCircuit(int[] gas, int[] cost){
        final int n = gas.length;
        if(n < 1)    return -1;

        int start = 0; // start station
        int i = 0; // passed station
        int tank = gas[0] - cost[0];

        while(true){
            while(tank < 0){
                --start;
                if(start == -1)    start += n;

                if(start == i)     return -1; // no station to move back
                tank += gas[start] - cost[start];
            }

            ++i;
            if(i == n)    i -= n;

            if(i == start)    break; // passed all stations
            tank += gas[i] - cost[i];
        }
        return start;
    }
}

/* unit test is in ../java_unittest/gasstation_junit */

File: leetcode/src/gasstation.cpp
/*
 * we have n gas statioin composing a loop.  gas[i] means the gas can be charged in station i, and cost[i] means the gas cost in the 
 * distance from [i] to [i+1](for [n-1], the next is [0])
 * assume our car has unlimited gas tank, ask whether can start from one station to run all of them
 *
 * test data:
 * gas{1,2,3,4,5}, cost{3,4,5,1,2}, n = 5, start = 3
 * */
#include "../include/preliminary.h"

class Solution{

public:
    int gasstation(vector<int>& gas, vector<int>& cost){
        const int n = gas.size();
        if(n < 1)    return -1;

        int start = 0; // start station
        int i = 0; // passed station
        int tank = gas[i] - cost[i];

        while(1){
            while(tank < 0){ // current start start station is not qualified, move start backward
                --start;
                if(start == -1)     start += n;

                if(start == i)    return -1; // no station to move back
                tank += gas[start] - cost[start];
            }

            ++i; // move forward
            if(i == n)    i -= n;

            if(i == start)    return start; // passed all stations
            tank += gas[i] - cost[i];
        }
        return -1;
    }
};

/* unit test is in ../cpp_unittest/gasstation_unittest */
